compiler/src/iree/compiler/Codegen/Common/FlattenMemRefs.cpp

```diff
@@ -164,185 +164,151 @@ static bool checkLayout(Value val) {
 }
 
 namespace {
-struct FlattenMemrefLoad : public OpRewritePattern<memref::LoadOp> {
-  using OpRewritePattern::OpRewritePattern;
 
-  LogicalResult matchAndRewrite(memref::LoadOp op,
+template <typename T>
+struct MemRefRewritePatternBase : public OpRewritePattern<T> {
+  using OpRewritePattern<T>::OpRewritePattern;
+  LogicalResult matchAndRewrite(T op,
                                 PatternRewriter &rewriter) const override {
-    Value memref = op.getMemref();
+    Value memref = getTargetMemref(op);
     if (!needFlattenning(memref) || !checkLayout(memref))
       return rewriter.notifyMatchFailure(op,
                                          "nothing to do or unsupported layout");
-
     auto &&[flatMemref, offset] = getFlattenMemrefAndOffset(
-        rewriter, op.getLoc(), memref, op.getIndices());
+        rewriter, op->getLoc(), memref, op.getIndices());
+    replaceOp(op, rewriter, flatMemref, offset);
+    return success();
+  }
+  virtual Value getTargetMemref(T op) const = 0;
+  virtual void replaceOp(Operation *op, PatternRewriter &rewriter,
```

Comment:
This could probably also take `T op`

And, looking around, it does confuse me why these are `virtual` - I see very little `virtual`-ing in the MLIR transforms, both IREE-side and upstream.

Usually, when there are repeated patterns like this, you just use a template argument for the operation, and then you might end up having

`template<> Value MemRefRewritePattern<memref::LoadOp> getTargetMemref() const`

for example.

Or, more commonly, you'd just have a `LoadLikePattern<memref::LoadOp>` `LoadLikePattern<vector::LoadOp>`, and so on, relying on the fact that various operations name their inputs the same.